Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -1633,6 +1635,25 @@ extension SourceKitLSPServer {
     return try await languageService.executeCommand(executeCommand)
   }
 
+  func getReferenceDocument(_ req: GetReferenceDocumentRequest) async throws -> GetReferenceDocumentResponse {
+    guard let givenURL = URL(string: req.uri.stringValue) else {
+      throw ResponseError.unknown("Invalid URL")
+    }
+
+    let filePath = "file:/" + givenURL.pathComponents.dropLast().joined(separator: "/")
+    let sourceFileURI = try DocumentURI(string: filePath)
+
+    guard let workspace = await workspaceForDocument(uri: sourceFileURI) else {
+      throw ResponseError.workspaceNotOpen(sourceFileURI)
+    }
+
+    guard let languageService = workspace.documentService.value[sourceFileURI] else {
+      throw ResponseError.unknown("No Language Service for URI: \(sourceFileURI)")
+    }
```

Comment:
You are really using implementation details of how the custom URL is formed to get the workspace for the document. We shouldn’t need to construct the `filePath`, instead we should call `workspaceForDocument(uri: req.uri)`. 

The way I would implement this is that `Workspace` gets a new method

```swift
/// Returns the primary document based on which the document at the given `uri` is based.
/// 
/// For macro expansions, this is the document from which the macro was expanded.
func primaryDocument(for uri: DocumentURI) -> DocumentURI {
  if fileURL.scheme == "swift-macro-expansion" {
    // your logic from lines 1643 and 1644.
  }
  return uri
}
```

And then in `SourceKitLSPServer.workspaceForDocument`, we use that primary file to call `fileHandlingCapability(for:)`.

That will give use the workspace that’s responsible for the macro expansion. To get the language service, we should wrap the current accesses to `workspace.documentService.value` in a function on `Workspace` that also calls `primaryDocument(for:)`. Something like the following (we can probably also mark `documentService` private then.

```swift
func documentService(for uri: DocumentURI) -> LanguageService? {
  return documentService.value[primaryDocument(for: uri)]
}
```